### PR TITLE
Annotation Idempotency

### PR DESCRIPTION
**Pull request by jihan-yin**
Adds `annotation_id` as optional parameter to annotations and prediction classes. This will be used to uniquely identify each annotation/prediction within its dataset item. Also adds `update` as an optional boolean parameter when uploading predictions and annotations. Setting `update=False`, the default behavior, will skip uploading annotations to dataset images that already have annotations with the same `annotation_id`. Setting `update=True` will overwrite such annotations.

### CODE REVIEW

File: nucleus/__init__.py
Comment (jihan-yin) [inline nucleus/__init__.py:131]: Was running into issues calling the api from within a notebook, this resolved it for me

File: nucleus/__init__.py
Comment (sasha-scale) [inline nucleus/__init__.py:465]: omg - nice fix

File: nucleus/__init__.py
Comment (sasha-scale) [inline nucleus/__init__.py:466]: see comment on sister PR - I think this should be optional with default value False

File: nucleus/__init__.py
Comment (sasha-scale) [inline nucleus/__init__.py:607]: remove print statement before merging

File: nucleus/constants.py
Comment (sasha-scale) [inline nucleus/constants.py:1]: remove before merging

File: nucleus/dataset.py
Comment (sasha-scale) [inline nucleus/dataset.py:19]: :nit: extraneous white space

**Review by sasha-scale — CHANGES_REQUESTED** [review]
I was unable to upload GT annotations without specifying the annotation_id, the error message was "annotation_id must be a string."  I think the issue was that when I did not specify it, it got sent as `None` in the request payload which didn't make it through schema validation

**Review by sasha-scale — APPROVED** [review]
LGTM, tested locally.  Response format does not have annotations_ignored which to me blocks merge, but that is more a comment on the scaleapi Pr than this one.

The unit tests are incredibly thorough!  great jobs!

File: tests/helpers.py
Comment (rkaplan) [inline tests/helpers.py:85]: Infinite recursion?

File: tests/helpers.py
Comment (rkaplan) [inline tests/helpers.py:89]: Infinite recursion?

File: tests/helpers.py
Comment (jihan-yin) [reply tests/helpers.py:85]: It's a different function 😄

File: tests/helpers.py
Comment (jihan-yin) [reply tests/helpers.py:89]: same as above

**Review by rkaplan — APPROVED** [review]

**Pull request merged**
